### DISCUSSION TITLE
Is it possible to have a ghost argument to a function?

### DISCUSSION THREAD

**Discussion by isubasinghe** [verifast/verifast#530] [Q&A] [answered]
Is there anyway to represent the following in verifast?

```c

int f(int a /*@, int b*@/) 
//@ requires b == 2;
//@ ensures true;
{

}

```

**Comment by isubasinghe** [comment]
I am aware that I can semantically have an equivalent through the use of structs but just wanted to check if this was at all possible?

**Comment by btj** [comment, marked as answer]
VeriFast does not currently support ghost parameters, but you can easily encode them by means of empty predicates such as `exists<T>` ([defined in prelude.h](https://github.com/verifast/verifast/blob/a514a79b30bd061a243befe1a787e54c234cc16c/bin/prelude.h#L13)), as follows:

```c
int f(int a) 
//@ requires exists<int>(?b) &*& b == 2;
//@ ensures true;
{
    //@ open exists(_);
}

// When calling f:
//@ close exists(2); // Pass 2 as value for `b` to `f`
f(10);
```

If you want to pass multiple ghost args, one option is to use pairs. For example, to pass three arguments with types T1, T2, T3, you can use `requires exists<pair<T1, pair<T2, T3> > >(pair(?v1, pair(?v2, ?v3))) &*& ...;`.

**Reply by isubasinghe** [reply]
Thanks a lot for your quick reply.
